gym_sts/test_valid_actions.py

```diff
@@ -13,44 +16,63 @@ def main():
     parser = argparse.ArgumentParser()
     parser.add_argument("lib_dir")
     parser.add_argument("mods_dir")
+    parser.add_argument("out_dir")
     parser.add_argument("--build_image", action="store_true")
     parser.add_argument("--headless", action="store_true")
     parser.add_argument("--runtime", default=30, type=int)
+    parser.add_argument("--allow_invalid", action="store_true")
+    parser.add_argument("--screenshots", action="store_true")
     args = parser.parse_args()
 
     if args.build_image:
         SlayTheSpireGymEnv.build_image()
-    env = SlayTheSpireGymEnv(args.lib_dir, args.mods_dir, headless=args.headless)
+    env = SlayTheSpireGymEnv(
+        args.lib_dir, args.mods_dir, args.out_dir, headless=args.headless
+    )
     env.reset(seed=42)
     rng = random.Random(42)
 
+    valid_choices = [True]
+    if args.allow_invalid:
+        valid_choices.append(False)
+
     num_steps = 0
     start_time = time.perf_counter()
 
     while True:
+        if args.screenshots:
+            env.screenshot(f"frame{num_steps:03d}.png")
+
         last_obs = env.observation_cache.get()
         assert last_obs is not None
+        print(last_obs.screen_type)
 
-        want_valid = rng.choice([True, False])
+        want_valid = rng.choice(valid_choices)
 
         actions = [
             action
             for action in ACTIONS
             if ActionValidators.validate(action, last_obs) == want_valid
         ]
 
+        if len(actions) == 0:
+            raise ValueError("No %svalid actions!" % ("" if want_valid else "in"))
+
         action = rng.choice(actions)
+        print(action)
         try:
             _, _, done, info = env.step(action._id)
         except TimeoutError as e:
             run_time = time.perf_counter() - start_time
             print(f"Error on step {num_steps} after {run_time} seconds.")
             print(e)
-            break
+            env.screenshot("error.png")
```

Comment:
Should this respect the value of `args.screenshots`? It might be surprising otherwise.

Reply:
I often want to screenshot on error and not every frame, so I think it's fine as-is.